### ISSUE TITLE
semantic: discrepancies in super diagnostics compared to TypeScript

### ISSUE THREAD

**Issue by ulrichstark** [oxc-project/oxc#13284]
In following code, OXC reports super 2, 3, and 4 and TypeScript reports super 1, 2, 3.

```typescript
class C {
   public fn() {
      super.superFn(); // 1, only TypeScript

      function sub() {
        super.superFn(); // 2, both
      }

      const obj = {
          fn1: function () {
            super.superFn(); // 3, both
          },
          fn2() {
            super.superFn(); // 4, only OXC
          },
      };
   }
}
```

[OXC Playground](
https://playground.oxc.rs/#eNqdkU9rwzAMxb+K8WmDMNbCLhk7bfS6sV17cVyleDhSkOxuJeS7T0matIWxw0LwH9nv955wZ70trY9OxDybbovGmDZXMXhT483tqaKf5Bb4bhw3evC4xfmkzuhTINQr1aXiN81U789iTyjJUPVpni6VSsVVeUZfc/9ij/zimrT+r7yf6ppXf1tYsmVnOeMwyRGT+7Zl4gyFjQHTvBZPLSybY1NRnHeJHUpN3NiydlGgL2zrWIAHoouRvt4hZcbXnCTsYHPqf5a3DHr3AG+OAWWqKmIwHxG6XhwmZnK8B01mQdb3qwftIQhFl2D3Avrq7Aa8kuY0nnawh7FBQFdF+KDMHhrXLm5NwFCH2U8fMDHFjUYfRAfgikS7n4BDIDV3kmz/A+F4w8k=)

[TypeScript Playground](
https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDC0G8BQ1UAcCuAjEBLY0AZgHYAUAlIiqjRBmgKYBOAdHY0wGJnkDcS1GkQzFgAF1wB7YtDpYKVIUPbM29Ztwr8l0AL4CdwaRDHRJWAFbQAvIp00SARgBcw0ROnQFye0pWt-TT5Be10AGhCdEgAmb0j7fzUOIO1fPQidXVS9JH0gA)

**Comment by Boshen** [comment]
These results are actually correct.

Oxc conforms to ECMA262, the ones reported by tsc but not present in Oxc are TypeScript specific errors.

**Comment by overlookmotel** [comment]
I believe one of the problems raised above is a genuine bug. Here's a reduced reproduction which is JS (not TS):

Parser says `super` is not legal in this position, but I believe it's wrong:

```js
class C {
  foo() {
    return {
      bar() {
        super.bar();
      },
    };
  }
}
```

[Playground](https://playground.oxc.rs/#eNpFUMFKxDAQ/ZUwJ4UiKnipR2Wvil73Mk2nSyDNlJlkdSn9d5OG1BDIm5fMe2+ygoUerEdV82bWczBmYr67r9AYoZgktMqYAeX/sixNC8nDTr82dusq2nZmO4e8oQOGfgVJoRx6CxF/oY+SqAPvQmxYLS90FLd5YN+qKBh0Ypmhn9ArbR0sKEpSFNF7/vna436kqG6kUwo2Os5+tX0Rym+v9IlCQSubJYr5LpHx4VA1I8qFcjIgfX58eskzOGWPkcZ3yn8mWOSzUktjeaQL7QNSwMHTNyexNONyuM0uuMk1P8shCvtTjl6ariQDa56+CpZA2Rw1wvYHy0aQKw==)

No errors if the object literal isn't inside a class method, so I assume there's some state about what context we're in which is getting confused. e.g. no errors for this:

```js
let _obj = {
  bar() {
    super.bar();
  },
};
```

[Playground](https://playground.oxc.rs/#eNo9j0FLxDAQhf9KmJNCERW8VLzJXhW9Lsg0O10iaabMTFaX0v++SUt7ypsh8733JvDQQiRzP9z9ujc3HZNzHcrd/Sqd0zySPCyr17qZm2Oai4IGGNoJJKf66DUZ/kNrkqmBGJJtWj2PtA/XoeO4TSaYtGcZoO0xKs0NjChKUokYI/99kWVJH9k0nOiQk7fAxW89H4XK3wt9olDSdVsQ1XxBFL07rExDOVNJBqTPj08vpUNQjmh0eicfUbDiC2lL4/lEZ1oKUsIu0jdn8TTguLsNIYU+bH6ekwnHQ4lejy4kHWtpvwJroGKOajDfABQ4hIE=)

**Comment by overlookmotel** [comment]
Sorry to open this again, but I still don't think we've got it quite right.

### 1. `super` in object method

This is legal in both JS and TS:

```js
class C {
  foo() {
    return {
      bar() {
        super.bar();
      },
    };
  }
}
```

[TS Playground](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDC0G8BQ1oDMD2GAUBKRKq0ATgKYAuArsQHYFFEBGYxe9DRElADqcQHTNWuANyEGAXwA046BLGoJSJUA)

### 2. `super` in class method

This is legal in JS, but TS have made it illegal:

```js
class C {
  foo() {
    super.foo();
  }
}
```

[TS Playground](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDC0G8BQ1oDMD2GAUBKRKq0EArgA4CmATgHSY64DchAvkm0A)

But it's only an error if the class doesn't extend a super-class. This does not produce an error in TS:

```js
class S {}

class C extends S {
  foo() {
    super.toString();
  }
}
```

[TS Playground](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEDK0G8C+AoFpIwMLQKYA8AXXAOwBMZ4EVpoAzAewYAoBKRG26CAVwAdcAJwB0hBrEKCAliQDmbANydUqIA)

In TS, I'm not sure if it should be considered an "early" syntax error, or a type-check error. TS [produces no error](https://www.typescriptlang.org/play/?target=99&noCheck=true#code/MYGwhgzhAEDC0G8BQ1oDMD2GAUBKRKq0EArgA4CmATgHSY64DchAvkm0A) with `noCheck: true`.

I think we can probably make it legal in both JS and TS.

### Where we're at

#13318 made (1) illegal in TS, but (2) legal. It should be the other way around.

### Edge cases

We probably also need to handle the cases which are fixed for object expressions in #13325 (but not fixed for classes).

**Comment by overlookmotel** [comment]
I feel ridiculous re-opening this *again*, but... #13323 did not fix this.

Semantic is currently flagging this use of `super` as illegal:

```js
class C {
  foo() {
    return {
      bar() {
        super.bar();
      },
    };
  }
}
```

[Oxc playground](https://playground.oxc.rs/#eNpFkMFOwzAMhl8l8gmkagIkLuUI2hUE113czJ2C0riyk8E09d1xUnVEkWI7jr//zxU89OAjqrpXdz0k50bmu/s1dE4oF0lb5tyA8n9Zl5aZZNfKL1t16dZoaZXlkGxDBwz9FaSkeuglZfyFPkuhDmJIeYvV80y35DINHLcsCyYdWSboR4xKSwczipLUiRgj/3w2ue8lazjSviSfAxuvdVuzkDWf6QOFkt7KykU87UOkhJOhIZPm3beCza/K2nyLb/gVmFFOZLKB9Onh8dkMBuWImY5vZB8qWNkbxZ57PtKJmnsDDZG+GnfCefVnLVNIYQwbz3PKwnFvvuqjM8nAavrWgVWQwVEzLH+Ffpos)

It is *not* illegal in either JS ([AST explorer](https://ast-explorer.dev/#eNpFjUEKAjEMRa8SslIYPMC49AriqptaI4x0mpK0ggy9u7FlMGTx8kLyN4w448u/vQZZcsEJswkfWJJxMA7Rq8IFNpcAnsyH40AAoVIl7RPA3ct/+SutmeTU9Xm3bRrUumkuWVsSW1K/dEhh9TcSXTg5nE1EX0iLw37pULlKoOsn01iv/KjR2F5h+wJBaz8Y)) or TS ([TS Playground](https://www.typescriptlang.org/play/?target=99#code/MYGwhgzhAEDC0G8BQ1oDMD2GAUBKRKq0ATgKYAuArsQHYFFEBGYxe9DRElADqcQHTNWuANyEGAXwA046BLGoJSJUA)).

#13323 fixed the behavior when not inside a class, but `super` in an object literal method *inside a class* is still broken.
